New prelude (#43)
RalfJung/minirust: prelude/main.md

# MiniRust prelude

Across all files in this repository, we assume some definitions to always be in scope.

```rust
/// All operations are fallible, so they return `Result`.  If they fail, that
/// means the program caused UB or put the machine to a halt.
pub type Result<T=()> = std::result::Result<T, TerminationInfo>;

#[non_exhaustive]
pub enum TerminationInfo {
  Ub(String),
  MachineStop(String),
}

> /// Some macros for convenient yeeting (yes this is valid syntax on nightly Rust).

Reviewer: ```suggestion
/// Some macros for convenient yeeting, i.e., return an error from a
/// `Option`/`Result`-returning function.
```
But also, we almost certainly don't want these macros in Coq like that.

Can you define a `yeet!` macro in specr so that these here can become `specr::yeet!(())` and `specr::yeet!(TerminationInfo::Ub(...))` etc? This will probably have to invoke the [FromResidual](https://doc.rust-lang.org/nightly/std/ops/trait.FromResidual.html) trait.

macro_rules! throw {
    ($($tt:tt)*) => {{
        None?;
        unreachable!()
    }};
}
macro_rules! throw_ub {
    ($($tt:tt)*) => {{
        Err(TerminationInfo::Ub(format!($($tt)*)))?;
        unreachable!()
    }};
}
macro_rules! throw_machine_stop {
    ($($tt:tt)*) => {{
        Err(TerminationInfo::MachineStop(format!($($tt)*)))?;
        unreachable!()
    }};
}

/// We leave the encoding of the non-determinism monad opaque.
pub use specr::Nondet;
pub type NdResult<T=()> = Nondet<Result<T>>;

/// Whether an integer value is signed or unsigned.
pub enum Signedness {
    Unsigned,
    Signed,
}
pub use Signedness::*;

/// Whether a pointer/reference/allocation is mutable or immutable.
pub enum Mutability {
    Mutable,
    Immutable,
}
pub use Mutability::*;

/// The type of mathematical integers.
/// We assume all the usual arithmetic operations to be defined.
pub use specr::BigInt;

> 
> impl BigInt {

Reviewer: ````suggestion
```

We document some relevant parts of the BigInt API here.

```rust,ignore
impl BigInt {
````

    /// Returns the unique value that is equal to `self` modulo `2^size.bits()`.
    /// If `signed == Unsigned`, the result is in the interval `0..2^size.bits()`,
    /// else it is in the interval `-2^(size.bits()-1) .. 2^(size.bits()-1)`.
    ///
    /// `size` must not be zero.

>     pub fn modulo(self, signed: Signedness, size: Size) -> BigInt { todo!() }

Reviewer: ```suggestion
    pub fn modulo(self, signed: Signedness, size: Size) -> BigInt;
```
